### PR TITLE
Add IDefaultCaseClause for default case clause

### PR DESCRIPTION
**Pull request by mavasani**
Fixes https://github.com/dotnet/roslyn/issues/20312

### CODE REVIEW

**Comment by heejaechang** [comment]
looks good.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 1]
```diff
@@ -926,16 +926,22 @@ private ISwitchStatement CreateBoundSwitchStatementOperation(BoundSwitchStatemen
...
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
-            return new LazySingleValueCaseClause(value, equality, caseKind, isInvalid, syntax, type, constantValue);
+            if (boundSwitchLabel.ExpressionOpt != null)
+            {
+                Lazy<IOperation> value = new Lazy<IOperation>(() => Create(boundSwitchLabel.ExpressionOpt));
+                BinaryOperationKind equality = GetLabelEqualityKind(boundSwitchLabel);
+                return new LazySingleValueCaseClause(value, equality, CaseKind.SingleValue, isInvalid, syntax, type, constantValue);
+            }
+            else
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: @gafter Is it correct to assume that any BoundSwitchLabel with no ExpressionOpt must be the "default" label?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 2]
```diff
@@ -1173,20 +1179,19 @@ private ISwitchStatement CreateBoundPatternSwitchStatementOperation(BoundPattern
...
             ITypeSymbol type = null;
             Optional<object> constantValue = default(Optional<object>);
 
             if (boundPatternSwitchLabel.Pattern.Kind == BoundKind.WildcardPattern)
             {
-                // Default switch label in pattern switch statement is represented as a regular case clause.
-                Lazy<IOperation> value = new Lazy<IOperation>(() => null);
-                return new LazySingleValueCaseClause(value, BinaryOperationKind.None, CaseKind.Default, isInvalid, syntax, type, constantValue);
+                // Default switch label in pattern switch statement is represented as a default case clause.
+                return new DefaultCaseClause( isInvalid, syntax, type, constantValue);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: I didn't find a test for this case (discard pattern being represented as a DefaultCaseClause).

Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Formatting nit: extra space in arguments.

Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: This is being covered here: https://github.com/dotnet/roslyn/pull/20707/files#diff-7d83c86cd85172860c4416b70dc793eaR270

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Note that this PR doesn't alter this logic. If this is an incorrect assumption, I can file a separate issue to track it.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 2 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs]: Ah. I got confused by the name. I thought this BoundKind was for discards, but it is actually for "default:". Thanks

---
In reply to: [126264567](https://github.com/dotnet/roslyn/pull/20707#discussion_r126264567) [](ancestors = 126264567)

**Review by jcouv — APPROVED** [review]
LGTM

File: src/Compilers/Core/Portable/Operations/OperationKind.cs [hunk 3]
```diff
@@ -199,5 +199,8 @@ public enum OperationKind
         DeclarationPattern = 0x410,
         /// <summary>Indicates an <see cref="IPatternCaseClause"/>.</summary>
         PatternCaseClause = 0x411,
+
+        /// <summary>Indicates an <see cref="IDefaultCaseClause"/>.</summary>
+        DefaultCaseClause = 0x4012,
```
Comment (cston) [inline src/Compilers/Core/Portable/Operations/OperationKind.cs]: `0x412`?

**Comment by cston** [comment]
LGTM, other than `OperationKind` value.

**Pull request merged**